docs/it/refactoring.md

```diff
@@ -0,0 +1,136 @@
+Sebbene questo potrebbe non essere sempre possibile, dopo avere estratto i metodi, ci potremmo trovare dinanzi a molti di questi che sono simili se non uguali. Con l'astrazione possiamo creare delle classi che contengono tali metodi. Metodi che potrebbero divenire ancora più semplici e generici, nonchè facili da interpretare.
+
+## Un Esempio
+
+Un caso semplice, ma spesso presente nel nostro codice, potrebbe essere quello di migliorare le performance rimuovendo le variabili temporanee (il codice che segue è pseudo-codice):
+
+```
+function applyDiscount() {
+  basePrice = cart.totalPrice();
+  return basePrice * discount();
+}
+```
+
+Il codice di partenza potrebbe già andare bene così ma, volendolo migliorare, potremmo dichiare la variabile temporanea in modo da facilitare al compilatore la vita, ovvero:
+
+```
+function applyDiscount() {
+  final basePrice = cart.totalPrice();
+  return basePrice * discount();
+}
+```
+
+Con questa modifica non abbiamo cambiato in alcun modo il comportamento. Ovviamente otterremo un piccolo vantaggio sulle performance, a patto che il linguaggio che stiamo usando possa trarre vantaggi dalla differenza tra variabili mutabili e non.
+Ma, riguardando ancora un attimo il codice, potremmo apportare un'altra modifica:
+
+```
+function applyDiscount() {
+  return cart.totalPrice() * discount();
+}
+```
+
+Rimuovendo la creazione di una variabile, che oggettivamente risulta solo superflua, andiamo a migliorare sicuramente le performance, non abbiamo introdotto alcuna complessità e la lettura del codice è ancora ampiamente comprensibile.
+
+## Costi
+
+Lo sviluppo di un software ha un costo, ovviamente anche il refactoring ne ha uno. Ma, fatto con regolarità, sfruttando tutte le occasioni possibili, ovvero effettuando micro-interventi di refactoring, non solo il suo costo è trascurabile ma, complessivamente, abbassa i costi grazie alla maggiore manutenibilità, scalabilità e performance ottenute.
+
+## Refactoring vs Riscrittura
+
+Si potrebbe obiettare che riscrivere una parte del nostro software, basandosi sull'esperienza acquisita, è più veloce o forse più auspicabile del refactoring.
+In realtà questa affermazione è fallace: riscrivere significa affrontare un'attività che comporta del tempo non sempre definito, che apporta delle modifiche al funzionamento del software, che andrà testato probabilmente con dei nuovi test e che, non da ultimo, andrà rilasciato in blocco.
+Con il refactoring, dove il dictat del non modificare in alcun modo il comportamento ci impone che i test devono continuare a funzionare senza alcun nostro intervento, è un'attività contingentata che può essere fatta durante lo sviluppo di una nuova funzionalità, se non durante la scrittura dei test stessi. E, soprattutto, il lavoro può essere rilasciato durante il normale ciclo di vita del software, annegato tra i vari rilasci e/o correzioni.
```

Comment:
Rivedrei questa frase che non mi suona benissimo

Reply:
Potrebbe sembrare contorta. La rileggo di nuovo e vedo se riesco a riformulare il concetto in maniera più chiara